app/classifier/tasks/survey/chooser/Choices.jsx

```diff
@@ -0,0 +1,183 @@
+import PropTypes from 'prop-types';
+import React from 'react';
+import sortIntoColumns from 'sort-into-columns';
+import Thumbnail from '../../../../components/thumbnail';
+
+const BACKSPACE = 8;
+const UP = 38;
+const DOWN = 40;
+
+class Choices extends React.Component {
+  constructor() {
+    super();
+    this.choiceButtons = [];
+  }
+
+  componentDidMount() {
+    this.sortChoiceButtons();
+  }
+
+  componentDidUpdate() {
+    this.sortChoiceButtons();
+  }
+
+  sortChoiceButtons() {
+    // overrides default DOM focus order by sorting the buttons according to task.choicesOrder
+    const newChoiceButtons = [];
+    this.choiceButtons
+      .filter(Boolean)
+      .map((button) => {
+        const choiceID = button.getAttribute('data-choiceID');
+        const index = this.props.task.choicesOrder.indexOf(choiceID);
+        newChoiceButtons[index] = button;
+      });
+    this.choiceButtons = newChoiceButtons.filter(Boolean);
+  }
+
+  howManyColumns({ length }) {
+    if (length <= 5) {
+      return 1;
+    } else if (length <= 20) {
+      return 2;
+    } else {
+      return 3;
+    }
+  }
+
+  whatSizeThumbnails({ length }) {
+    if (length <= 5) {
+      return 'large';
+    } else if (length <= 10) {
+      return 'medium';
+    } else if (length <= 20) {
+      return 'small';
+    } else if (this.props.task.alwaysShowThumbnails) {
+      return 'small';
+    } else {
+      return 'none';
+    }
+  }
+
+  handleKeyDown(choiceId, e) {
+    const index = this.choiceButtons.indexOf(document.activeElement);
+    let newIndex;
+    switch (e.which) {
+      case BACKSPACE:
+        this.props.onRemove(choiceId);
+        e.preventDefault();
+        break;
+      case UP:
+        newIndex = index - 1;
+        if (newIndex === -1) {
+          newIndex = this.choiceButtons.length - 1;
+        }
+        this.choiceButtons[newIndex].focus();
+        e.preventDefault();
+        break;
+      case DOWN:
+        newIndex = (index + 1) % this.choiceButtons.length;
+        this.choiceButtons[newIndex].focus();
+        e.preventDefault();
+        break;
+      default:
+    }
+  }
+
+  render() {
+    const { filteredChoices, task, translation } = this.props;
+    this.choiceButtons = [];
```

Comment:
I know this was in the original code, but it's confusing to me to have this reset to an empty array on render. Not a blocker for this, but something to consider for future refactors.

Reply:
If you don’t reset your refs on each render, the array will just grow and grow and grow. There’s probably a better way to do this.